src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,135 @@
+                fileChangeWatcher,
+                globalOptionService,
+                loggerFactory,
+                listenerProvider,
+                projectLoadTelemetry,
+                serverConfigurationFactory,
+                binlogNamer)
+    {
+        _lspServices = lspServices;
+        _logger = loggerFactory.CreateLogger<FileBasedProgramsProjectSystem>();
+        _metadataAsSourceFileService = metadataAsSourceFileService;
+    }
+
+    public Workspace Workspace => ProjectFactory.Workspace;
+
+    public async Task<TextDocument?> AddMiscellaneousDocumentAsync(Uri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var documentPath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+        var container = documentText.Container;
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentPath, container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var metadataDoc = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return metadataDoc;
+        }
+
+        var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentPath}");
+            return null;
+        }
+
+        // For Razor files we need to override the language name to C# as thats what code is generated
+        var isRazor = languageInformation.LanguageName == "Razor";
+        var languageName = isRazor ? LanguageNames.CSharp : languageInformation.LanguageName;
+        var loadedProject = await CreateAndTrackInitialProjectAsync(documentPath, language: languageName);
+        var documentFileInfo = new DocumentFileInfo(documentPath, logicalPath: documentPath, isLinked: false, isGenerated: false, folders: default);
+        var projectFileInfo = new ProjectFileInfo()
+        {
+            Language = languageName,
+            FilePath = uri.IsFile ? VirtualProject.GetVirtualProjectPath(documentPath) : documentPath,
+            CommandLineArgs = uri.IsFile ? ["/langversion:preview", "/features:FileBasedProgram=true"] : ["/langversion:preview"],
+            Documents = isRazor ? [] : [documentFileInfo],
+            AdditionalDocuments = isRazor ? [documentFileInfo] : [],
+            AnalyzerConfigDocuments = [],
+            ProjectReferences = [],
+            PackageReferences = [],
+            ProjectCapabilities = [],
+            ContentFilePaths = [],
+            FileGlobs = []
+        };
+        await loadedProject.UpdateWithNewProjectInfoAsync(projectFileInfo, _logger);
```

Comment:
We still need to remember to set the "HasAllInformation" flag here.

Reply:
Considering the following scenarios:
1) ordinary project. HasAllInformation should always be true, when this method runs, I think.
2) loose file, not part of a file-based program. HasAllInformation should be false.
3) file-based program entry point (currently). HasAllInformation should be true.

So I was wondering if we should pass in a flag like `forMiscellaneousFiles` when creating the `LoadedProject`, and finally say `HasAllInformation = !forMiscellaneousFiles || checkHeuristic()` within this method (since it's FBP-ness can change from edit-to-edit.)

Reply:
Done